torch/distributed/_tensor/_utils.py

```diff
@@ -0,0 +1,66 @@
+from typing import Tuple, Sequence
+from torch.distributed._tensor.device_mesh import DeviceMesh
+from torch.distributed._tensor.placement_types import Shard, Placement
+from torch._prims_common import ShapeType
+
+
+def compute_local_shape(
+        global_shape: ShapeType,
+        mesh: DeviceMesh,
+        placements: Sequence[Placement]
+    ) -> Tuple[int, ...]:
+    """
+    Compute the shape of a local shard of the given DTensor on its current
+    coordinate of the mesh.
+    """
+    local_shape = global_shape  # start with global shape
+    ndim = len(global_shape)
+    for idx, placement in enumerate(placements):
+        mesh_dim_size = mesh.size(idx)
+        my_coordinate = mesh.get_coordinate()
+        assert my_coordinate is not None, "Rank not part of mesh!"
+        if isinstance(placement, Shard):
+            shard_dim = placement.dim
+            assert (
+                shard_dim < ndim
+            ), f"Sharding dim {shard_dim} greater than tensor ndim {ndim}"
+            local_shard_size, _ = placement._local_shard_size_on_dim(
+                local_shape[shard_dim], mesh_dim_size, my_coordinate[idx]
+            )
```

Comment:
Nit: might further consolidate the duplicate logic?

Reply:
Yeah we can see how to consolidate the logic further, I was thinking if I should merge these two APIs, but since they are for different purposes, we can separate them for now.